packages/@react-spectrum/s2/src/Table.tsx

```diff
@@ -11,85 +11,1089 @@
+    setIsInResizeMode(false);
+    propsOnResizeEnd?.(widths);
+  }, [propsOnResizeEnd, setIsInResizeMode]);
+
+  let context = useMemo(() => ({
+    isQuiet,
+    density,
+    overflowMode,
+    selectionStyle,
+    loadingState,
+    isInResizeMode,
+    setIsInResizeMode,
+    scale,
+    layout
+  }), [isQuiet, density, overflowMode, selectionStyle, loadingState, scale, layout, isInResizeMode, setIsInResizeMode]);
+
+  let isLoading = loadingState === 'loading' || loadingState === 'loadingMore';
+  let scrollRef = useRef(null);
+  let memoedLoadMoreProps = useMemo(() => ({
+    isLoading: isLoading,
+    onLoadMore
+  }), [isLoading, onLoadMore]);
+  useLoadMore(memoedLoadMoreProps, scrollRef);
+
+  return (
+    <ResizableTableContainer
+      onResize={propsOnResize}
+      onResizeEnd={onResizeEnd}
+      onResizeStart={onResizeStart}
+      className={(UNSAFE_className || '') + mergeStyles(tableWrapper, styles)}
+      style={UNSAFE_style}>
+      <UNSTABLE_Virtualizer layout={layout}>
+        <InternalTableContext.Provider value={context}>
+          <RACTable
+            ref={scrollRef}
+            className={renderProps => table({
+              ...renderProps,
+              isQuiet
+            })}
+            selectionBehavior={selectionStyle === 'highlight' ? 'replace' : 'toggle'}
+            onRowAction={onAction}
+            {...otherProps} />
+        </InternalTableContext.Provider>
+      </UNSTABLE_Virtualizer>
+    </ResizableTableContainer>
+  );
+}
+
+const centeredWrapper = style({
+  display: 'flex',
+  alignItems: 'center',
+  justifyContent: 'center',
+  width: 'full',
+  height: 'full'
+});
+
+export interface TableBodyProps<T> extends Omit<RACTableBodyProps<T>, 'style' | 'className' | 'dependencies'> {}
+
+export function TableBody<T extends object>(props: TableBodyProps<T>) {
+  let {items, renderEmptyState, children} = props;
+  let tableVisualOptions = useContext(InternalTableContext);
+  let {loadingState} = tableVisualOptions;
+  let emptyRender;
+  let renderer = children;
+  let stringFormatter = useLocalizedStringFormatter(intlMessages, '@react-spectrum/s2');
+  let loadMoreSpinner = (
+    <UNSTABLE_TableLoadingIndicator className={style({height: 'full', width: 'full'})}>
+      <div className={centeredWrapper}>
+        <ProgressCircle
+          isIndeterminate
+          aria-label={stringFormatter.format('table.loadingMore')} />
+      </div>
+    </UNSTABLE_TableLoadingIndicator>
+  );
+
+  // If the user is rendering their rows in dynamic fashion, wrap their render function in Collection so we can inject
+  // the loader. Otherwise it is a static renderer and thus we can simply add the table loader after
+  // TODO: this assumes that the user isn't providing their children in some wrapper though and/or isn't doing a map of children
+  // (though I guess they wouldn't provide items then so the check for this is still valid in the latter case)...
+  if (typeof children === 'function' && items) {
```

Comment:
will need to think a bit more about this and how best to tell if someone is providing a render function or just mapping over items directly